File: str_helper.cpp
// system includes
#include <map>

// includes
#include "str_helper.h"
#include "basic_parser/str_helper.h"

namespace generic_protocol
{

namespace str_helper
{

using ::basic_parser::str_helper::write;
using ::basic_parser::str_helper::write_t;

// enums

#define TUPLE_VAL_STR(_x_)  _x_,#_x_

std::ostream & write( std::ostream & os, const ErrorResponse_type_e r )
{
    typedef ErrorResponse_type_e Type;
    static const std::map< Type, std::string > m =
    {
        { Type:: TUPLE_VAL_STR( AUTHENTICATION_ERROR ) },
        { Type:: TUPLE_VAL_STR( NOT_PERMITTED ) },
        { Type:: TUPLE_VAL_STR( INVALID_ARGUMENT ) },
        { Type:: TUPLE_VAL_STR( RUNTIME_ERROR ) },
    };

    auto it = m.find( r );

    static const std::string undef( "undef" );

    if( it != m.end() )
        return write( os, it->second );

    return write( os, undef );
}

// objects

std::ostream & write( std::ostream & os, const SessionInfo & r )
{
    os << "(";

    os << " user_id="; write( os, r.user_id );
    os << " start_time="; write( os, r.start_time );
    os << " expiration_time="; write( os, r.expiration_time );

    os << ")";

    return os;
}

// base messages

std::ostream & write( std::ostream & os, const ForwardMessage & r )
{
    os << "(";


    os << ")";

    return os;
}

std::ostream & write( std::ostream & os, const BackwardMessage & r )
{
    os << "(";


    os << ")";

    return os;
}

std::ostream & write( std::ostream & os, const Request & r )
{
    os << "(";

    os << " session_id="; write( os, r.session_id );

    os << ")";

    return os;
}

// messages

std::ostream & write( std::ostream & os, const ErrorResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );

    os << " type="; write( os, r.type );
    os << " descr="; write( os, r.descr );

    return os;
}

std::ostream & write( std::ostream & os, const AuthenticateRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const ForwardMessage&>( r ) );

    os << " user_login="; write( os, r.user_login );
    os << " password="; write( os, r.password );

    return os;
}

std::ostream & write( std::ostream & os, const AuthenticateAltRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const ForwardMessage&>( r ) );

    os << " user_id="; write( os, r.user_id );
    os << " password="; write( os, r.password );

    return os;
}

std::ostream & write( std::ostream & os, const AuthenticateResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );

    os << " session_id="; write( os, r.session_id );

    return os;
}

std::ostream & write( std::ostream & os, const CloseSessionRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const ForwardMessage&>( r ) );

    os << " session_id="; write( os, r.session_id );

    return os;
}

std::ostream & write( std::ostream & os, const CloseSessionResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );


    return os;
}

std::ostream & write( std::ostream & os, const GetUserIdRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const Request&>( r ) );

    os << " user_login="; write( os, r.user_login );

    return os;
}

std::ostream & write( std::ostream & os, const GetUserIdResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );

    os << " user_id="; write( os, r.user_id );

    return os;
}

std::ostream & write( std::ostream & os, const GetSessionInfoRequest & r )
{
    // base class
    str_helper::write( os, static_cast<const Request&>( r ) );

    os << " id="; write( os, r.id );

    return os;
}

std::ostream & write( std::ostream & os, const GetSessionInfoResponse & r )
{
    // base class
    str_helper::write( os, static_cast<const BackwardMessage&>( r ) );

    os << " session_info="; write( os, r.session_info );

    return os;
}

} // namespace str_helper

} // namespace generic_protocol

File: str_helper.php
<?php

namespace generic_protocol;


// includes
require_once __DIR__.'/../basic_parser/str_helper.php';

// enums

function to_string__ErrorResponse_type_e( $r )
{
    $map = array
    (
        ErrorResponse_type_e__INVALID_OR_EXPIRED_SESSION => 'INVALID_OR_EXPIRED_SESSION',
        ErrorResponse_type_e__NOT_PERMITTED => 'NOT_PERMITTED',
        ErrorResponse_type_e__INVALID_ARGUMENT => 'INVALID_ARGUMENT',
        ErrorResponse_type_e__RUNTIME_ERROR => 'RUNTIME_ERROR',
    );

    if( array_key_exists( $r, $map ) )
    {
        return $map[ $r ];
    }

    return '?';
}

// objects

function to_string__SessionInfo( & $r )
{
    $res = "";

    $res .= "(";

    $res .= " user_id=" . \basic_parser\to_string__int( $r->user_id );
    $res .= " start_time=" . \basic_parser\to_string__int( $r->start_time );
    $res .= " expiration_time=" . \basic_parser\to_string__int( $r->expiration_time );

    $res .= ")";

    return $res;
}

// base messages

function to_string__ForwardMessage( & $r )
{
    $res = "";

    $res .= "(";

    // no base class

    $res .= ")";

    return $res;
}

function to_string__BackwardMessage( & $r )
{
    $res = "";

    $res .= "(";

    // no base class

    $res .= ")";

    return $res;
}

function to_string__Request( & $r )
{
    $res = "";

    $res .= "(";

    // base class
    $res .= to_string__ForwardMessage( $r );

    $res .= " session_id=" . \basic_parser\to_string__string( $r->session_id );

    $res .= ")";

    return $res;
}

// messages

function to_string__ErrorResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );

    $res .= " type=" . to_string__ErrorResponse_type_e( $r->type );
    $res .= " descr=" . \basic_parser\to_string__string( $r->descr );

    return $res;
}

function to_string__AuthenticateRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__ForwardMessage( $r );

    $res .= " user_login=" . \basic_parser\to_string__string( $r->user_login );
    $res .= " password=" . \basic_parser\to_string__string( $r->password );

    return $res;
}

function to_string__AuthenticateAltRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__ForwardMessage( $r );

    $res .= " user_id=" . \basic_parser\to_string__int( $r->user_id );
    $res .= " password=" . \basic_parser\to_string__string( $r->password );

    return $res;
}

function to_string__AuthenticateResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );

    $res .= " session_id=" . \basic_parser\to_string__string( $r->session_id );

    return $res;
}

function to_string__CloseSessionRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__ForwardMessage( $r );

    $res .= " session_id=" . \basic_parser\to_string__string( $r->session_id );

    return $res;
}

function to_string__CloseSessionResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );


    return $res;
}

function to_string__GetUserIdRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__Request( $r );

    $res .= " user_login=" . \basic_parser\to_string__string( $r->user_login );

    return $res;
}

function to_string__GetUserIdResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );

    $res .= " user_id=" . \basic_parser\to_string__int( $r->user_id );

    return $res;
}

function to_string__GetSessionInfoRequest( & $r )
{
    $res = "";

    // base class
    $res .= to_string__Request( $r );

    $res .= " id=" . \basic_parser\to_string__string( $r->id );

    return $res;
}

function to_string__GetSessionInfoResponse( & $r )
{
    $res = "";

    // base class
    $res .= to_string__BackwardMessage( $r );

    $res .= " session_info=" . to_string__SessionInfo( $r->session_info );

    return $res;
}

// generic

function to_string( $obj )
{
    $handler_map = array(
        // objects
        'generic_protocol\SessionInfo'         => 'to_string__SessionInfo',
        // messages
        'generic_protocol\ErrorResponse'         => 'to_string__ErrorResponse',
        'generic_protocol\AuthenticateRequest'         => 'to_string__AuthenticateRequest',
        'generic_protocol\AuthenticateAltRequest'         => 'to_string__AuthenticateAltRequest',
        'generic_protocol\AuthenticateResponse'         => 'to_string__AuthenticateResponse',
        'generic_protocol\CloseSessionRequest'         => 'to_string__CloseSessionRequest',
        'generic_protocol\CloseSessionResponse'         => 'to_string__CloseSessionResponse',
        'generic_protocol\GetUserIdRequest'         => 'to_string__GetUserIdRequest',
        'generic_protocol\GetUserIdResponse'         => 'to_string__GetUserIdResponse',
        'generic_protocol\GetSessionInfoRequest'         => 'to_string__GetSessionInfoRequest',
        'generic_protocol\GetSessionInfoResponse'         => 'to_string__GetSessionInfoResponse',
    );

    $type = get_class( $obj );

    if( array_key_exists( $type, $handler_map ) )
    {
        $func = '\\generic_protocol\\' . $handler_map[ $type ];
        return $func( $obj );
    }

    return NULL;
}

# namespace_end generic_protocol


?>
